components/scream/src/physics/shoc/shoc_update_prognostics_implicit_impl.hpp

```diff
@@ -13,14 +13,159 @@ namespace shoc {
+  const auto last_nlevi_indx = (nlevi-1)%Spack::n;
+
+  // linearly interpolate tkh, tk, and air density onto the interface grids
+  linear_interp(team,zt_grid,zi_grid,tkh,tkh_zi,nlev,nlevi,0);
+  linear_interp(team,zt_grid,zi_grid,tk,tk_zi,nlev,nlevi,0);
+  linear_interp(team,zt_grid,zi_grid,rho_zt,rho_zi,nlev,nlevi,0);
+
+  // Define the tmpi variable, which is really dt*(g*rho)**2/dp
+  // at interfaces. Substitue dp = g*rho*dz in the above equation
+  compute_tmpi(team, nlevi, dtime, rho_zi, dz_zi, tmpi);
+
+  // compute 1/dp term, needed in diffusion solver
+  dp_inverse(team, nlev, rho_zt, dz_zt, rdp_zt);
+  team.team_barrier();
+
+  // compute terms needed for the implicit surface stress (ksrf)
+  // and tke flux calc (wtke_sfc)
+  Scalar ksrf, wtke_sfc;
+  {
+    const auto wsmin = 1;
+    const auto ksrfmin = 1e-4;
+    const auto ustarmin = 0.01;
+
+    const auto rho = rho_zi(last_nlevi_pack)[last_nlevi_indx];
+    const auto uw = uw_sfc;
+    const auto vw = vw_sfc;
+
+    const auto taux = rho*uw;
+    const auto tauy = rho*vw;
+
+    const auto u_wind_sfc = u_wind(last_nlev_pack)[last_nlev_indx];
+    const auto v_wind_sfc = v_wind(last_nlev_pack)[last_nlev_indx];
+
+    const auto ws = ekat::impl::max(std::sqrt((u_wind_sfc*u_wind_sfc) + v_wind_sfc*v_wind_sfc), sp(wsmin));
+    const auto tau = std::sqrt(taux*taux + tauy*tauy);
+    ksrf = ekat::impl::max(tau/ws, sp(ksrfmin));
+
+    const auto ustar = ekat::impl::max(std::sqrt(std::sqrt(uw*uw + vw*vw)), sp(ustarmin));
+    wtke_sfc = ustar*ustar*ustar;
+  }
+
+  // compute surface fluxes for liq. potential temp, water and tke
+  {
+    const auto cmnfac = dtime*(C::gravit*rho_zi(last_nlevi_pack)[last_nlevi_indx]*rdp_zt(last_nlev_pack)[last_nlev_indx]);
+    Kokkos::single(Kokkos::PerTeam(team), [&] () {
+      thetal(last_nlev_pack)[last_nlev_indx] += cmnfac*wthl_sfc;
+      qw(last_nlev_pack)[last_nlev_indx] += cmnfac*wqw_sfc;
+      tke(last_nlev_pack)[last_nlev_indx] += cmnfac*wtke_sfc;
+    });
+
+    const auto tracer_sfc = Kokkos::subview(tracer, nlev-1, Kokkos::ALL());
+    const Int num_tracer_pack = ekat::npack<Spack>(num_tracer);
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(team, num_tracer_pack), [&] (const Int& p) {
+      tracer_sfc(p) += cmnfac*wtracer_sfc(p);
+    });
+  }
+
+  // Store RHS values in X1 and X2
+  team.team_barrier();
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team, nlev), [&] (const Int& k) {
+    const auto k_view_indx = k/Spack::n;
+    const auto k_pack_indx = k%Spack::n;
+
+    X1(k,0)[0] = u_wind(k_view_indx)[k_pack_indx];
+    X1(k,1/Spack::n)[1%Spack::n] = v_wind(k_view_indx)[k_pack_indx];
+
+    X2(k,0)[0] = thetal(k_view_indx)[k_pack_indx];
+    X2(k,1/Spack::n)[1%Spack::n] = qw(k_view_indx)[k_pack_indx];
+    X2(k,2/Spack::n)[2%Spack::n] = tke(k_view_indx)[k_pack_indx];
+    for (Int p=0; p<num_tracer; ++p) {
+      X2(k,(3+p)/Spack::n)[(3+p)%Spack::n] = tracer(k, p/Spack::n)[p%Spack::n];
```

Comment:
Here you load the tracer array into another array. This is one of ~3 approaches to the problem, and I'll argue roughly the worst of the 3. Here are the approaches:
1. Tracers come in at the highest level with fast index k. Load them into a transposed tracer array for SHOC, with an extra 3 slots for other RHS. Use that array as the RHS; here, there would be no copy into another array. That's the approach I would take b/c it would save copying/transposes in the case of multiple SHOC iterations. Small win b/c we're expecting only one iteration in practice.
2. Tracers come in at the highest level with fast index k. Do nothing at the high level. Here, do a transpose at the same time as you copy. An acceptable alternative to 1.
3. Tracers come in at the highest level with fast index k. Load them into a transposed tracer array for SHOC. Then, here, copy to another array. This has needless copying. This is what's being done here.
So I'd like you to redesign this part to use approaches 1 or 2 or anything else you think of that's at least as good as either of those.

I really like how you've reduced this code to exactly two solves, the minimum number possible.

Separate point: Note here and in any tracer transpose code you end up writing that `scalarize`ing the standard arrays before this ||4 would save all the pack indexing arithmetic. You'd just write `X1(k,1/Spack::n)[1%Spack::n] = v_wind[k]`, for example.